librz/include/rz_analysis.h

```diff
@@ -1435,7 +1435,54 @@ RZ_API void rz_analysis_esil_interrupts_fini(RzAnalysisEsil *esil);
 RZ_API void rz_analysis_esil_mem_ro(RzAnalysisEsil *esil, int mem_readonly);
 RZ_API void rz_analysis_esil_stats(RzAnalysisEsil *esil, int enable);
 
-/* trace */
+/* new trace implementation */
+typedef enum {
+	TRACE_READ, ///< read
+	TRACE_WRITE ///< write
+} TRACE_OPERATION;
+
+typedef struct {
+	ut64 addr; ///< memory address
+	TRACE_OPERATION behavior; ///< read or write, see TRACE_OPERATION enums
+	ut64 value; ///< data either written to or read from
+} RzILTraceMemOp;
+
+typedef struct {
+	char *reg_name; ///< name of register
+	TRACE_OPERATION behavior; ///< READ or WRITE, see TRACE_OPERATION enums
+	ut64 value; ///< data either written to or read from
+} RzILTraceRegOp;
+
+enum {
+	TRACE_INS_HAS_MEM_R = 0x1U, ///< instruction include memory read
+	TRACE_INS_HAS_MEM_W = 0x2U, ///< instruction include memory write
+	TRACE_INS_HAS_REG_R = 0x4U, ///< instruction include register read
+	TRACE_INS_HAS_REG_W = 0x8U ///< instruction include register write
+};
+
+typedef struct {
+	ut64 addr; ///< Address of instruction
+	ut32 stats; ///< Has write/read to reg/mem ? see TRACE_INS_HAS_* enums
+
+	RzPVector *write_mem_ops; ///< Vector<RzILTraceRegOp>
+	RzPVector *read_mem_ops; ///< Vector<RzILTraceRegOp>
+
+	RzPVector *write_reg_ops; ///< Vector<RzILTraceRegOp>
+	RzPVector *read_reg_ops; ///< Vector<RzILTraceRegOp>
```

Comment:
@ret2libc too much memory allocations, don't you think? `RzILTraceInstruction` -> `RzPVector` -> `RzILTraceRegOp`/`RzILTraceMemOp`